### PR TITLE
ARROW-16778: [C++] Fix win32 build and tests

### PR DESCRIPTION
**Pull request by jduo**

### CODE REVIEW

**Comment by jduo** [comment]
> Looks like it still doesn't build on 32-bit Windows and there are tons of compiler warnings about casts (one of which is an error). We might want a general way to deal with `int64_t -> size_t` conversions.

I'm thinking to add a templated ToSizeT function to a utils header for converting integer types to size_t, that adds pragmas to suppress warnings on 32-bit platforms (or when going from a signed integer).

**Comment by lidavidm** [comment]
FWIW, an explicit static_cast is enough to suppress the warning. I'm more thinking: on 32-bit platforms, do we also want to check for potential overflow?

File: .github/workflows/cpp.yml [hunk 1]
```diff
@@ -197,12 +197,21 @@ jobs:
...
       matrix:
-        os:
-          - windows-2019
         include:
           - os: windows-2019
-            name: Windows 2019
+            name: Windows 2019 x64
             generator: Visual Studio 16 2019
+            platform: x64
+            openssl-n-bits: 64
```
Comment (pitrou) [inline .github/workflows/cpp.yml:205]: Hmm, what is this for? Can I find some documentation for this somewhere? A Google search yields nothing.

File: cpp/src/arrow/util/io_util.cc [hunk 2]
```diff
@@ -1439,7 +1439,8 @@ Status MemoryAdviseWillNeed(const std::vector<MemoryRegion>& regions) {
     PrefetchEntry(const MemoryRegion& region)  // NOLINT runtime/explicit
         : VirtualAddress(region.addr), NumberOfBytes(region.size) {}
   };
-  using PrefetchVirtualMemoryFunc = BOOL (*)(HANDLE, ULONG_PTR, PrefetchEntry*, ULONG);
+  using PrefetchVirtualMemoryFunc =
+      BOOL(__stdcall*)(HANDLE, ULONG_PTR, PrefetchEntry*, ULONG);
```
Comment (pitrou) [inline cpp/src/arrow/util/io_util.cc:1443]: Ha :-)

File: .github/workflows/cpp.yml [hunk 3]
```diff
@@ -197,12 +197,21 @@ jobs:
...
+            cmake-args: -A x64 -DOPENSSL_ROOT_DIR="C:\Program Files\OpenSSL-Win64"
+            chocolate-force:
+          - os: windows-2019
+            name: Windows 2019 Win32
```
Comment (pitrou) [inline .github/workflows/cpp.yml:209]: Win32 is the former name of the Windows API
```suggestion
            name: Windows 2019 x86
```

**Comment by pitrou** [comment]
I would be fine with `static_cast<size_t>`. A 32-bit build cannot represent any in-memory containers larger than 2**32 elements (except perhaps a BooleanArray...).

**Comment by pitrou** [comment]
ChunkedArray is still in-memory. For other places (such as file access) we might want to check.

File: .github/workflows/cpp.yml (hunk 1 above)
Comment (jduo) [reply .github/workflows/cpp.yml:205]: These are build variables in the GitHub workflow.
Platform is used to specify the CPU configuration to build using MSBuild.
openssl-n-bits is used to expand the directory name that OpenSSL got installed to using the chocalatey package manager.

Comment (pitrou) [reply .github/workflows/cpp.yml:205]: I understand about `platform`, but I can't find any reference to `openssl-n-bits` anywhere. Did I miss something?

Comment (pitrou) [reply .github/workflows/cpp.yml:205]: but `${{ matrix.openssl-n-bits }}` is never used below, so how does it work?

**Comment by jduo** [comment]
@lidavidm  @pitrou , the warnings are now all fixed for the 32-bit MSVC build.

It is failing to build linking some parquet tests that depend on boost::filesystem (trying to link to the 64-bit version incorrectly). I did set ARROW_PARQUET to off in the github workflow though, so I'm not sure why it's even trying to build this test or the parquet code.

This build is also failing archery linting. Mostly around line wrapping. Is there an automated way of fixing the archery formatting warnings?

**Comment by lidavidm** [comment]
@jduo you can use Archery for autoformatting: https://arrow.apache.org/docs/developers/cpp/development.html#code-style-linting-and-ci

**Comment by jduo** [comment]
@github-actions autotune

**Comment by lidavidm** [comment]
@jduo can you rebase here? The autotune, weirdly enough, does not actually kick off CI.

**Comment by lidavidm** [comment]
@jduo it turns out ARROW_SUBSTRAIT implies ARROW_PARQUET so we should turn that off, too

**Comment by lidavidm** [comment]
Hmm, I suppose you could still have a NA array with > INT_MAX rows on 32-bit, since there's no actual storage…

**Comment by lidavidm** [comment]
Or actually. A boolean array could also technically exceed size_t rows, since there's 8 rows per byte?

**Review by lidavidm — COMMENTED** [review]
I scanned through things and things generally look good, though are we worried about NA/boolean arrays? If so, some kernels and I/O may need to do some overflow checks. 

Also, have we tested what happens if we try to read a very large IPC file on 32-bit Windows?

**Comment by pitrou** [comment]
While it is technically ok, I think sprinkling `static_cast<size_t>` everywhere is both cumbersome (on the code writing _and_ reading sides) and will turn out fragile maintenance-wise.

I also don't think we're very interested in ensuring highest quality support for 32-bit Windows, so I would favor an approach that's much lighter on maintenance and code writing demands.
(for example by selectively disabling some warnings on 32-bit MSVC)

**Comment by jduo** [comment]
> While it is technically ok, I think sprinkling `static_cast<size_t>` everywhere is both cumbersome (on the code writing _and_ reading sides) and will turn out fragile maintenance-wise.
> 
> I also don't think we're very interested in ensuring highest quality support for 32-bit Windows, so I would favor an approach that's much lighter on maintenance and code writing demands. (for example by selectively disabling some warnings on 32-bit MSVC)

The argument against disabling compilation warnings is that if the warnings come from a public header, it the library user inherits the warning and can't really work around it. They would need to disable the warning on their side if they report warnings as failures in their build system, which may be undesirable for example.

**Comment by pitrou** [comment]
Well, we _could_ selectively disable warnings in Arrow headers, but I fear that a lot of headers would probably be affected :-(

Perhaps this needs to be further discussed on the mailing-list. My opinion is that 32-bit support should not come with undue maintenance load.

**Comment by amol-** [comment]
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍

**Pull request closed without merging**